contracts/data/buyAndBurn/SwapAndAddLiquidityLib.sol

```diff
@@ -0,0 +1,353 @@
+                    params.tokenOut,
+                    result.tokenInUnused,
+                    result.tokenOutReceived,
+                    params.lpTokenId,
+                    params.positionManager
+                );
+
+            result.lpAdded = lpAmount;
+
+            // Calculate spare amounts
+            if (params.tokenIn < params.tokenOut) {
+                // tokenIn is token0
+                result.tokenInUnused -= amount0Used;
+                result.tokenOutSpare = result.tokenOutReceived - amount1Used;
+            } else {
+                // tokenIn is token1
+                result.tokenInUnused -= amount1Used;
+                result.tokenOutSpare = result.tokenOutReceived - amount0Used;
+            }
+
+            emit LiquidityAdded(
+                params.tokenIn < params.tokenOut ? params.tokenIn : params.tokenOut,
+                params.tokenIn < params.tokenOut ? params.tokenOut : params.tokenIn,
+                amount0Used,
+                amount1Used,
+                uint128(lpAmount),
+                params.lpTokenId
+            );
+        } else {
+            // No LP position or no tokens to add - all tokenOut is spare
+            result.tokenOutSpare = result.tokenOutReceived;
+        }
+
+        return result;
+    }
+
+    /**
+     * @notice Quote the best swap amount within price impact threshold
+     * @dev Calculates maximum swap amount that stays within impact threshold
+     */
+    function _quoteBestSwap(
+        address tokenIn,
+        address tokenOut,
+        uint256 amountIn,
+        uint256 impactThreshold,
+        uint24 poolFee,
+        address factory
+    ) private view returns (
+        uint256 amountToSwap,
+        uint256 expectedOut,
+        uint256 priceImpact
+    ) {
+        // Get pool
+        address pool = IUniswapV3Factory(factory).getPool(tokenIn, tokenOut, poolFee);
+        if (pool == address(0)) {
+            return (0, 0, 10000); // 100% impact if no pool
+        }
+
+        // Get pool reserves (simplified - in production, use sqrtPriceX96)
+        (uint160 sqrtPriceX96,,,,,,) = IUniswapV3Pool(pool).slot0();
+        uint128 liquidity = IUniswapV3Pool(pool).liquidity();
+
+        if (liquidity == 0) {
+            return (0, 0, 10000);
+        }
+
+        // Calculate price impact for full amount
+        // This is simplified - production should use Uniswap's quoter
+        uint256 fullImpact = _estimatePriceImpact(amountIn, liquidity);
+
+        if (fullImpact <= impactThreshold) {
+            // Can swap full amount
+            amountToSwap = amountIn;
+            expectedOut = _estimateSwapOutput(amountIn, sqrtPriceX96, liquidity);
+            priceImpact = fullImpact;
+        } else {
+            // Binary search for maximum amount within threshold
+            uint256 left = 0;
+            uint256 right = amountIn;
+
+            while (right - left > amountIn / 1000) { // 0.1% precision
```

Comment:
I don't think we need a binary search for best swap here because the strategy is that we should swap to get as much as possible tokenOut for both LP and burn (see Miro buy-and-burn diagram). It's different from the strategy we used in DLP rewards, in which we need to find the best swap to maximize added liquidity and minimize the amount of spare tokens (it's not a goal in buy-and-burn).

I also don't see quote and / or swap with `singleBatchImpactThreshold` and `perSwapSlippageCap` implemented.